### DISCUSSION TITLE
Fluttering Z height when Arc Welder is on (first layer)

### DISCUSSION THREAD

**Discussion by Dizzybird** [FormerLurker/ArcWelderPlugin#252] [Q&A] [answered]
Hello, I'm sorry if this has already been addressed, but I was not able to find a solution yet. I have Arc Welder installed on Cura slicer as well I've used it on Octoprint. The issue I am running into is Z height is moving and/or fluttering up and down, but only for the first layer. I can see the bed moving up and down by very small increments when it is printing the first layer. I have tried it with the Cura plugin as well on the Octoprint plugin. I don't have this issue unless I turn on Arc Welding feature. If I disable it, the issue goes away.
Any thoughts on why this is happening? I'm all ears and thank you for any response one may have to assist!

My printer is an Ender 5 plus with updated Marlin Firmware
I only have this issue with Arc Welding on
I do NOT have Z hop on or any other feature that would cause this (that I know of)
First layer is acceptable at best but ultimately have been turning Arc welder off and then issue goes away

**Comment by seantapscott** [comment]
The only reason that the Z would be adjusting on the first layer would be
the mesh bed settings you have set.

Read more here:
https://marlinfw.org/docs/gcode/G029-ubl.html

And M420 also has more info:
https://marlinfw.org/docs/gcode/M420.html

I have the exact same printer.  The z-mesh bed levelling fade height will
also stop adjusting Z higher up on the model (generally 2mm is the
default).

Likely nothing to do with arcwelder, except that you may have mesh bed
levelling turned on in the start gcode in Cura when you are slicing using
arcwelder. .


- Just a fellow Arcwelder user

**Reply by Dizzybird** [reply]
Edit:  This is in my start Gcode in Cura
 G28 ;Home
M420 S1 Z2 ;Enable ABL using saved Mesh and Fade Height

Thank you for the reply. I will look into this more.

 If I send a file that I know has no issues printing (and no first layer Z height movements) to Octoprint, when I compress said file in Octoprint with ArcWelder. Only then does that file now have this issue with Z height moving all over the place with the first layer (and I mean ALL over the place. First layer is .2mm. After the first layer all is back to normal. If I hadn't compressed the same file in Octoprint with Arcwelder, it would NOT have this Z height movement on first layer. The same scenario would occur if I had used Arcwelder in Cura instead of Octoprint. These are prints files I use all the time. Something about engaging Arcwelder is making this happen, or causing it to conflict with an otherwise normal functioning print.
Sorry for the long winded description.  I've had these ArcWelder plugins for awhile, just now only starting to test using it.

Thanks for the info and links, in the meantime I will do some more troubleshooting

**Comment by Dizzybird** [comment]
If anyone has any ideas on why this would be happening, I would love to know the answer. I can't seem to find much info on ArcWelder troubleshooting for Z height issues.
I'm really at a loss, and the ArcWelder plugin is unusable for me with this issue
Thoughts?
Thank you!

**Reply by seantapscott** [reply]
Post your gcode.

**Reply by Dizzybird** [reply]
I posted in the comment above my Cura start Gcode for the bed leveling. Here is the whole start Gcode in Cura 

M201 X500.00 Y500.00 Z100.00 E5000.00 ;Setup machine max acceleration
M203 X500.00 Y500.00 Z10.00 E50.00 ;Setup machine max feedrate
M204 P500.00 R1000.00 T500.00 ;Setup Print/Retract/Travel acceleration
M205 X8.00 Y8.00 Z0.40 E5.00 ;Setup Jerk
M220 S100 ;Reset Feedrate
M221 S100 ;Reset Flowrate

G28 ;Home
M420 S1 Z2 ;Enable ABL using saved Mesh and Fade Height

G92 E0 ;Reset Extruder
G1 Z2.0 F3000 ;Move Z Axis up
G1 X350 Y340 Z0.28 F5000.0 ;Move to start position
G1 X350 Y60.0 Z0.28 F1500.0 E15 ;Draw the first line

G92 E0 ;Reset Extruder
G1 Z2.0 F3000 ;Move Z Axis up

**Reply by seantapscott** [reply]
Cool.  And is there a file where you declare that that the printer isn't flying around all around?

**Reply by Dizzybird** [reply]
None of my files that I print do this until I engage ArcWelder. It's only on the first layer when Z level is moving all over. These prints have a small footprint, set in the center of the bed. I've never seen this before until now. 
I don't have any recent changes that I've made to the printer, that I can think would cause this?
Thanks for helping with the troubleshooting

**Reply by seantapscott** [reply]
What kind of changes did you make to the firmware?  Increase command length?  Do you have dynamic precision on in ArcWelder?  And did you enable Arcs?  M115 outputs "arcs: 1" in the list?

**Reply by Dizzybird** [reply]
Ah ha! I think you found it. To be honest, I had just left the Arc Welder settings it installed with. But I see that "G90 Influences the Extruder" was checked in the settings. I assume that is the culprit. It says it should be checked for Marlin 2.0 users (which I am not).

I will look at the Arcwelder settings more closely. Do you have any recommendations on settings that I should tweak? 
Otherwise, I think the problem is solved. Thanks for helping. Greatly appreciated!

**Reply by Dizzybird** [reply]
To answer your question about these settings. I did not change any firmware settings or command length. I do not have Dynamic Precision on. The only Enable Arcs I see is "enable travel arcs", and it is not on. Where do I enable Arcs, and M115 outputs "arcs: 1" in the list?

**Comment by seantapscott** [comment, marked as answer]
You have to enable arcs when you compile firmware. If you can see console
output from your printer, command M115 's report should show arcs as a
value of one if it was compiled as arcs.

Previously, the developer of ArcWelderLib and I were examining strange
behaviors with arc commands on Creality printers.  We came to understand it
as caused by command truncation. The stock firmware could only hold like 60
character command in its buffer.  Since arc commands were longer
(especially with dynamic precision), the command buffer would overrun,
causing partial truncation, and the first 3-10 characters would be
dropped.  A third person suggested a new Marlin setting in the nightly
builds that specifically targeted this issue.

So we used the nightly build from that point in time to compile new
firmware for our ender printers that would enable long arcs. I also
increased the communication buffer to like 1023 bytes and turned on Serial
XON/XOFF or something.   Without reexamining my own github history on
arcwelder, I don't remember the dates of that.

Boosted the firmware line processing speed from like 125 lines per second
with 3% retries to 420 lines per second with 0.1% retries in testing
(RepetierServer offers this communication testing functionality).

Anyway, overall, arcwelder is really just a gcode compression algorithm,
which offers very little improvements in print quality unless you're
dealing with print stuttering or if you just really want more
human-readable gcode.

**Reply by Dizzybird** [reply]
Thanks for your assistance in trying to solve this. I thought I had it figured out, but for some reason it is still causing me problems. As you said Arcwelder is going to offer very little improvements visible on the print quality in my circumstance, and since everything else is working and printing fine. I will return to this when I have more time to troubleshoot and test some prints. 
Thanks again!
